sklearn/_build_utils/__init__.py

```diff
@@ -33,3 +35,42 @@ def atlas_not_found(blas_info_):
         cblas_libs = blas_info.pop('libraries', [])
 
     return cblas_libs, blas_info
+
+
+def get_cython_source(filename):
+    is_c_filename = filename.endswith('.c')
+    is_cpp_filename = filename.endswith('.cpp')
+
+    # files in src are .c and .cpp files that are not cython-generated
+    if ('src' + os.sep) in filename and (is_c_filename or is_cpp_filename):
+        return filename
+    elif is_c_filename:
+        filename = filename[:-1]
+    elif is_cpp_filename:
+        filename = filename[:-3]
+    else:
+        raise ValueError('Only .c and .cpp files are supported. '
+                         'Got {0!r} instead'.format(filename))
+    return filename + 'pyx'
+
+
+def add_cython_extension(top_path, config, name, sources, **kwargs):
+    is_dev_version = not os.path.exists(os.path.join(top_path, 'PKG-INFO'))
+
+    if is_dev_version:
+        sources = [get_cython_source(filename) for filename in sources]
```

Comment:
This is the part that switches between using .c{,pp} files when building a release and the .pyx when building a dev version.